### PR TITLE
Feat/implement tax number

### PR DESCRIPTION
**Pull request by ILoveMyTortoise**

### CODE REVIEW

File: src/Nox.Reference.Data.World/Extensions/Queries/TaxNumberQueryExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    /// <param name="validationNumber">Tax number as string</param>
+    /// <param name="shouldValidateViaApi">Flag to determine if validation should use online API service (if applicable) or not</param>
+    /// <returns>Validation result</returns>
+    public static TaxNumberValidationResult? Validate(
+        this IQueryable<TaxNumberDefinition> query,
+        string countryAlpha2Code,
+        string validationNumber,
+        bool shouldValidateViaApi = true)
+    {
+        var definition = query.FirstOrDefault(x => x.Country.AlphaCode2.ToUpper() == countryAlpha2Code.ToUpper());
```
Comment (AndreyDegtyarev) [inline src/Nox.Reference.Data.World/Extensions/Queries/TaxNumberQueryExtensions.cs:56]: potential NullReferenceException here

File: src/Nox.Reference.Data.World/Extensions/Queries/TaxNumberQueryExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    /// <param name="query">Current collection</param>
+    /// <param name="countryAlpha2Code">Alpha 2 country code. Example: "UA".</param>
+    /// <param name="validationNumber">Tax number as string</param>
+    /// <param name="shouldValidateViaApi">Flag to determine if validation should use online API service (if applicable) or not</param>
+    /// <returns>Validation result</returns>
+    public static TaxNumberValidationResult? Validate(
+        this IQueryable<TaxNumberDefinition> query,
+        string countryAlpha2Code,
+        string validationNumber,
+        bool shouldValidateViaApi = true)
```
Comment (AndreyDegtyarev) [inline src/Nox.Reference.Data.World/Extensions/Queries/TaxNumberQueryExtensions.cs:54]: Can we avoid using default parameter shouldValidateViaApi  and  introduce overload method.

File: src/Nox.Reference.Data.World/Configurations/TaxNumbers/TaxNumberDefinitionConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+{
+    protected override void DoConfigure(EntityTypeBuilder<TaxNumberDefinition> builder)
+    {
+        builder
+            .HasMany(x => x.ValidationRules)
+            .WithOne();
+
+        builder
+            .HasOne(x => x.Country)
+            .WithOne(x => x.TaxNumberDefinition);
```
Comment (AndreyDegtyarev) [inline src/Nox.Reference.Data.World/Configurations/TaxNumbers/TaxNumberDefinitionConfiguration.cs:16]: Foreign key configuration should appear here

File: src/Nox.Reference.Data.World/Models/VatNumbers/ValidationStatus.cs [hunk 4]
```diff
@@ -1,6 +1,6 @@
 ﻿namespace Nox.Reference;
 
-public enum VatValidationStatus
+public enum ValidationStatus
```
Comment (AndreyDegtyarev) [inline src/Nox.Reference.Data.World/Models/VatNumbers/ValidationStatus.cs:3]: May be leave as it was. ValidationStatus is more generic name for all validations

File: src/Nox.Reference.Data.World/Models/TaxNumbers/TaxNumberValidationResult.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    public void AddError(string error)
+    {
+        if (error != null)
+        {
+            Status = ValidationStatus.Invalid;
+        }
+        _validationErrors.Add(error);
+    }
+
+    public void AddErrors(IEnumerable<string> errors)
```
Comment (AndreyDegtyarev) [inline src/Nox.Reference.Data.World/Models/TaxNumbers/TaxNumberValidationResult.cs:53]: result.AddErrors(ChecksumValidationHelper.ValidateLuhnDigitNumber(digitPart));

Looks like we expect error here. Might well method name be changed. 
AddErrors -> HandleValidation() or Check() etc..

Comment (AndreyDegtyarev) [reply src/Nox.Reference.Data.World/Models/TaxNumbers/TaxNumberValidationResult.cs:53]: Also, maybe signature can be changed Func<string, string>() validationFn ?

File: tests/Nox.Reference.Data.World.Tests/TaxNumberTests.cs [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+            // not found, in root
+            if (path.Parent == null)
+            {
+                path = new DirectoryInfo(Directory.GetCurrentDirectory());
+                break;
+            }
+            path = path.Parent;
+        }
+        // TODO: come up with different test data
+        _testFilePath = Path.Combine(path.FullName, "data/tests/VatNumbers/");
```
Comment (AndreyDegtyarev) [inline tests/Nox.Reference.Data.World.Tests/TaxNumberTests.cs:45]: can we move it to configuration file?

**Review by AndreyDegtyarev — APPROVED** [review]

File: src/Nox.Reference.Data.World/Extensions/Queries/TaxNumberQueryExtensions.cs (hunk 1 above)
Comment (ILoveMyTortoise) [reply src/Nox.Reference.Data.World/Extensions/Queries/TaxNumberQueryExtensions.cs:56]: Handled by nullable flow in the project

File: src/Nox.Reference.Data.World/Entities/Countries/Country.cs [hunk 7]
```diff
@@ -28,10 +28,13 @@ public class Country : NoxReferenceEntityBase,
     public CountryCapital Capital => Capitals.FirstOrDefault() ?? new CountryCapital();
 
     internal int? VatNumberDefinitionId { get; private set; }
-
     [ForeignKey("VatNumberDefinitionId")]
     public virtual VatNumberDefinition? VatNumberDefinition { get; internal set; }
 
+    internal int? TaxNumberDefinitionId { get; private set; }
+    [ForeignKey("TaxNumberDefinitionId")]
```
[the remark this answers is not in the document]
Comment (ILoveMyTortoise) [reply src/Nox.Reference.Data.World/Entities/Countries/Country.cs:35]: With annotation we can only specify field as a string name, not as a field, so it's a worse option

File: src/Nox.Reference.Data.World/Extensions/Queries/TaxNumberQueryExtensions.cs (hunk 2 above)
Comment (ILoveMyTortoise) [reply src/Nox.Reference.Data.World/Extensions/Queries/TaxNumberQueryExtensions.cs:54]: removed default value

File: src/Nox.Reference.Data.World/Configurations/TaxNumbers/TaxNumberDefinitionConfiguration.cs (hunk 3 above)
Comment (ILoveMyTortoise) [reply src/Nox.Reference.Data.World/Configurations/TaxNumbers/TaxNumberDefinitionConfiguration.cs:16]: answered above

File: src/Nox.Reference.Data.World/Models/VatNumbers/ValidationStatus.cs (hunk 4 above)
Comment (ILoveMyTortoise) [reply src/Nox.Reference.Data.World/Models/VatNumbers/ValidationStatus.cs:3]: It's a generic enum for validations in scope of this project

File: src/Nox.Reference.Data.World/Configurations/TaxNumbers/TaxNumberValidationRuleConfiguration.cs [hunk 8]
```diff
@@ -0,0 +1,11 @@
+﻿using Microsoft.EntityFrameworkCore.Metadata.Builders;
+
+namespace Nox.Reference.Data.World.Configurations.VatNumbers;
+
+internal class TaxNumberValidationRuleConfiguration : NoxReferenceEntityConfigurationBase<TaxNumberValidationRule>
+{
+    protected override void DoConfigure(EntityTypeBuilder<TaxNumberValidationRule> builder)
+    {
+        builder.OwnsOne(x => x.Checksum);
```
[the remark this answers is not in the document]
Comment (ILoveMyTortoise) [reply src/Nox.Reference.Data.World/Configurations/TaxNumbers/TaxNumberValidationRuleConfiguration.cs:9]: No, can't be deleted, it works and changes structure

File: src/Nox.Reference.Data.World/Models/TaxNumbers/TaxNumberValidationResult.cs (hunk 5 above)
Comment (ILoveMyTortoise) [reply src/Nox.Reference.Data.World/Models/TaxNumbers/TaxNumberValidationResult.cs:53]: Renamed

File: tests/Nox.Reference.Data.World.Tests/TaxNumberTests.cs (hunk 6 above)
Comment (ILoveMyTortoise) [reply tests/Nox.Reference.Data.World.Tests/TaxNumberTests.cs]: done

**Pull request merged**